packages/salesforcedx-vscode-apex/src/commands/forceApexExecute.ts

```diff
@@ -0,0 +1,232 @@
+      .withFlag('--apexcodefile', data.fileName)
+      .withLogName('force_apex_execute')
+      .build();
+  }
+}
+
+interface ApexExecuteParameters {
+  apexCode?: string;
+  fileName?: string;
+}
+
+export class AnonApexGatherer
+  implements ParametersGatherer<ApexExecuteParameters> {
+  public async gather(): Promise<
+    CancelResponse | ContinueResponse<ApexExecuteParameters>
+  > {
+    if (hasRootWorkspace()) {
+      const editor = vscode.window.activeTextEditor;
+      if (!editor) {
+        return { type: 'CANCEL' };
+      }
+
+      const document = editor.document;
+      if (!editor.selection.isEmpty || document.isUntitled) {
+        return {
+          type: 'CONTINUE',
+          data: {
+            apexCode: !editor.selection.isEmpty
+              ? document.getText(editor.selection)
+              : document.getText()
+          }
+        };
+      }
+
+      return { type: 'CONTINUE', data: { fileName: document.uri.fsPath } };
+    }
+    return { type: 'CANCEL' };
+  }
+}
+
+export class ApexLibraryExecuteExecutor extends LibraryCommandletExecutor<
+  ApexExecuteParameters
+> {
+  public static diagnostics = vscode.languages.createDiagnosticCollection(
+    'apex-errors'
+  );
+
+  constructor() {
+    super(
+      nls.localize('apex_execute_text'),
+      'force_apex_execute_library',
+      OUTPUT_CHANNEL
+    );
+  }
+
+  public async run(
+    response: ContinueResponse<ApexExecuteParameters>
+  ): Promise<boolean> {
+    const connection = await workspaceContext.getConnection();
+    const executeService = new ExecuteService(connection);
+    const { apexCode, fileName: apexFilePath } = response.data;
+
+    const result = await executeService.executeAnonymous({
+      apexFilePath,
+      apexCode
+    });
+
+    this.outputResult(result);
+
+    const editor = vscode.window.activeTextEditor;
+    const document = editor!.document;
+    const filePath = apexFilePath ?? document.uri.fsPath;
+
+    this.handleDiagnostics(result, filePath);
+
+    return result.success;
+  }
+
+  private outputResult(response: ExecuteAnonymousResponse): void {
+    let outputText = '';
+    if (response.success) {
+      outputText += `${nls.localize('apex_execute_compile_success')}\n`;
+      outputText += `${nls.localize('apex_execute_runtime_success')}\n`;
+      outputText += `\n${response.logs}`;
+    } else {
+      const diagnostic = response.diagnostic![0];
+
+      if (!response.compiled) {
+        outputText += `Error: Line: ${diagnostic.lineNumber}, Column: ${diagnostic.columnNumber}\n`;
+        outputText += `Error: ${diagnostic.compileProblem}\n`;
+      } else {
+        outputText += `${nls.localize('apex_execute_compile_success')}\n`;
+        outputText += `Error: ${diagnostic.exceptionMessage}\n`;
+        outputText += `Error: ${diagnostic.exceptionStackTrace}\n`;
+        outputText += `\n${response.logs}`;
+      }
+    }
+    OUTPUT_CHANNEL.appendLine(outputText);
+  }
+
+  private handleDiagnostics(
```

Comment:
Took the opportunity to make the result formatting and diagnostics logic private methods of this executor because they are internal details of the command. More encapsulated than have them scattered in separate files.

Reply:
Cool, as long as this doesn't break testing since sometimes we've left things as public so they can be tested.